### PR TITLE
[IndexTable] Allow row click when onClick is passed in on non-selectable tables

### LINKED ISSUE #11518: onClick Disabled on IndexTable.Row when IndexTable prop 'selectable' is set to false
## Issue summary

Setting the `IndexTable` prop of selectable to `false` causes the onClick callback of `IndexTable.Row` to not fire. 

## Expected behavior

I would assume that you would still be able to have an onClick event on the row even if the checkboxes are disabled/hidden. I'm on the fence as to whether this is intentional or not as the cursor is set to default in this state.

In the below example I would expect handleClick() to still fire.

```jsx
<IndexTable selectable={false}>
  <IndexTable.Row onClick={() => handleClick()}>
        {/* row content */}
  </IndexTable.Row>
</IndexTable>
```

### PR DESCRIPTION
**Pull request by sydturn**
### WHY are these changes introduced?

Resolves https://github.com/Shopify/polaris/issues/11518 and Relates to https://github.com/Shopify/credit/issues/1252 which needs to use clickable IndexTables but without checkboxes.

Here is the [Figma](https://www.figma.com/file/LtoBWJzLY17yfm9fYxRHrD/Finances-unification?type=design&node-id=4145-35105&mode=design&t=0WNE8MDZa62DwdMj-0) I'm working off of. Currently these tables exist in Shopify Balance and Credit already but they're using our own custom css/html alongside ResourceList and it's
a) hard to manage (the columns don't have the context of the other columns and don't automatically adjust so we have to hardcode the column widths which is a pain whenever we need to add a new column
b) doesn't conform to Polaris aesthetics
So we're [trying to migrate](https://github.com/Shopify/credit/issues/1252) to using IndexTables to make maintenance on these tables easier, the transition to the new Figma easier, and the aesthetics of the tables look more like Polaris.
I have it [mostly implemented ](https://github.com/Shopify/web/pull/120937)but the only thing is that the rows need to be clickable without the checkboxes.

There are no aesthetic changes in this PR for including a screenshot.

### WHAT is this pull request doing?

This pull requests makes it so that if an `onClick` event is provided to a `IndexTable.Row` that is in an `IndexTable` that has `selectable={false}` the `onClick` will still be called. Rows being selectable vs. clickable should not be mutually exclusive. 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#install-dependencies-and-build-workspaces)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

☝️ TLDR; 
1. `git clone https://github.com/Shopify/polaris.git`
2. `yarn install && cd polaris-react`
3. `dev up && dev server`

I have provided a code snippet that you can put in your playground for testing. Where in `main` clicking a row would do nothing, with the changes in this PR clicking on a row will log to the console. 

```
    <Page title="Playground">
      <IndexTable
        itemCount={2}
        selectable={false}
        headings={[{title: 'date'}, {title: 'description'}, {title: 'amount'}]}
      >
        <IndexTable.Row onClick={handleOnClick} position={1} id="1">
          <IndexTable.Cell>Today</IndexTable.Cell>
          <IndexTable.Cell>Dollarama</IndexTable.Cell>
          <IndexTable.Cell>$30</IndexTable.Cell>
        </IndexTable.Row>
      </IndexTable>
    </Page>
```

You can also test it on [this](https://admin.web.banking-credit-f7h4.sydney-rhude.us.spin.dev/store/shop3/credit?referrer=finances-nav&modalView=receiptView) spin instance (the transaction table shown utilizes non-selectable but yes clickable)

### 🎩 checklist

I'm not so sure this needs to be added to the changeset, documentation, or readme, as it is just implementing a fix for something that should already exist but doesn't? I'm not sure. This isn't a new feature, it's just a fix for an existing one that rarely gets used I guess?

- [x] Tested a [snapshot](https://github.com/Shopify/polaris/blob/main/documentation/Releasing.md#-snapshot-releases)
- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by etlgfx** [comment]
Oh I fixed this too here: https://github.com/Shopify/polaris/pull/11652 but feel free to ship this if you have time to dedicate to it.

I took a slightly different approach, but also found a user created issue that I linked in my description.
(I'm not on Polaris team)

**Comment by sydturn** [comment]
> Oh I fixed this too here: #11652 but feel free to ship this if you have time to dedicate to it.
> 
> I took a slightly different approach, but also found a user created issue that I linked in my description. (I'm not on Polaris team)

Oh nice! Yours makes a lot more sense, of course we would want to skip all the other "selectable" functionality if `selectable={false}` and only do the `onClick` stuff 🤦‍♀️ 

I added the extra check to mine too 😄 good callout!

File: .changeset/odd-swans-teach.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+---
+---
```
Comment (chloerice) [inline .changeset/odd-swans-teach.md]:
```suggestion
--- 
'@shopify/polaris': patch
---

Fixed `IndexTable.Row` `onClick` not being called when `selectable` is `false`
```

**Review by chloerice — APPROVED** [review]
Thanks for contributing @sydturn! Changes 👀 👍🏽

File: polaris-react/src/components/IndexTable/components/Row/Row.tsx [hunk 2]
```diff
@@ -120,13 +120,14 @@ export const Row = memo(function Row({
...
     !selectable &&
+      !onClick &&
       !primaryLinkElement.current &&
       styles['TableRow-unclickable'],
   );
 
   let handleRowClick;
 
-  if ((!disabled && selectable) || primaryLinkElement.current) {
+  if ((!disabled && selectable) || onClick || primaryLinkElement.current) {
```
Comment (etlgfx) [inline polaris-react/src/components/IndexTable/components/Row/Row.tsx:130]: I was hoping we could simplify this to basically only check `disabled`, all the extra boolean logic is confusing and could lead to weird edge cases, thoughts?

**Comment by etlgfx** [comment]
@sydturn Looks great so far, thanks for taking this on!!

File: polaris-react/src/components/IndexTable/components/Row/Row.tsx (hunk 2 above)
Comment (sydturn) [reply polaris-react/src/components/IndexTable/components/Row/Row.tsx:130]: I do worry a lot about changing the existing functionality too much, I'd rather leave it as designed 😅 I figure the extensive unit tests should be handling the edge cases... though I suppose they should also alert us to any breaking changes we'd make by changing this...

Edit: So it seems like the functionality is that we don't want them to be able to click the row at all, and moving those checks down makes it so the row is clickable again ...I think I'd rather leave it as is if that's fine by you (yay for unit tests 👍 )

Comment (sydturn) [reply polaris-react/src/components/IndexTable/components/Row/Row.tsx:130]: Actually, for added context: When removing the other checks and moving it to the nested if statement the unit test that fails is this one: 

```
it('does not fire onClick handler when row is clicked and no primary link child present and table is not selectable', () => {
    const row = mountWithTable(
      <Row {...defaultProps}>
        <th>
          <a href="/">Child without data-primary-link</a>
        </th>
      </Row>,
      {
        indexTableProps: {
          itemCount: 1,
          selectable: false,
        },
      },
    );

    const onClick = () => {
      row.find(Row)!.find('tr')!.trigger('onClick');
    };

    expect(onClick).toThrow(
      'Attempted to call prop onClick but it was not defined.',
    );
  });
```
The expected substring is `'Attempted to call prop onClick but it was not defined.'` but it gets `"Cannot read properties of undefined (reading 'stopPropagation')"`. I noticed most of the other tests call `triggerOnClick(row, 1, defaultEvent);` instead of finding the row and clicking on it but doing so makes it so the click doesn't throw at all. Even when I change the code in `Row` back to the original though and call `triggerOnClick` in that unit test instead it still doesn't throw. 

Anyway the TLDR; is that I'm worried about changing these existing unit tests to fit the new code rather than just ensuring my new code doesn't break existing unit tests. I just don't have the proper context I think on the nuances of Polaris to be futzing around with it all so I'd like to just leave it as is.

**Review by itwasmattgregg — APPROVED** [review]
Looks great!
